### ISSUE TITLE
Define the infrastructure and build process

### ISSUE THREAD

**Issue by christophetd** [turingwars/turingwars#5]
Properties we want: 

- Infrastructure as a code and configuration management, we want to be able to easily migrate server without loosing everything. Suggestions: Ansible / Bash scripts / Terraform / Docker
- Easy deployment pipeline, e.g. just a bash script

Suggestion:

- Terraform to define our base "infrastructure" (which at this point should be a single VM). Essentially it will allow us to spawn a VM in a cloud we want (e.g. DigitalOcean), with what we need preinstalled on it
-  Bash scripts or Python for deployment, e.g. to build the game engine, scp the application code + the JAR to the server, and restart the application
- For frontend: ? (@hmil) but I'm guessing something like nginx and PM2

**Comment by christophetd** [comment]
Another possibility could be to use Ansible for both provisioning and deployment. I did this for another project, and found it easy to use.

The general flow is:

```
# Once-time VM provisioning
$ ansible-playbook -i ansible/provisioning/main.yml
[connects to the machine over SSH]
[installs Java]
[installs nginx and configuration files]
[installes node and npm]

# Deployment
$ ansible-playbook -i ansible/deployment/main.yml
[clones latest repo]
[rebuilds latest engine JAR]
[restarts the application]
```

**Comment by hmil** [comment]
I suggest:
- Do a release on GH which includes a pre-built zip archive which contains:
  - the server code
  - the built client code (web/public)
  - the VM jar
  - none of the tests or unnecessary sources
  - a shell script ( + batch script for win ?) which contains the one-liner: `NODE_ENV=production cd web && node dist/server.js`
  - Maybe script the archive generation process so we don't f*ck it up doing it manually.
- Assert the prerequisites: Node.JS >= 8 and JVM >= 8

For deployment, just spawn a VM, install java and nodejs, DL the archive and spin it up. Maybe setup monit or pm2 and a firewall if you want to get fancy.

That's simple enough. There should be no infrastructure code in this repo IMO so if you want to script the deployment or use a fancy infra tool that can live in another repo.

**Comment by hmil** [comment]
Also there are some tools out there which lets Travis push the github release when it sees a particular branch or tag.

**Comment by christophetd** [comment]
Yeah that's probably easy to achieve, we could also use the `TRAVIS_BRANCH` environment variable in order to run a deployment if the current branch is `master`

**Comment by hmil** [comment]
bikeshed detail: `master` should really be what you mean by `dev` branch because PRs always target master on github. Instead, have a `release` or `stable` branch for releases.

**Comment by christophetd** [comment]
Or maybe work with tags as you suggested, could make the process easier. We can just [create a new release](https://github.com/turingwars/turingwars/releases/new) on Github's interface, or `git tag v1.0 && git push --tags`
